### ISSUE TITLE
1D Wave Simulation froze my computer

### ISSUE THREAD

**Issue by VictorSeven** [godotengine/godot#21894]
Hi there,

I was trying to do a simple visualization of a 1D wave in Godot. Until now I had 0 problems with the editor. However, this time, my entire computer froze several times when running the app. Once my screen went corrupted, so I am pretty sure that the problem is that the graphic card froze. 
After a while with my computer entirely frozen, it recovers and closes Godot. I am sure it is the engine because it was the only thing opened and it always happened each time I ran the project.

[I upload here a version of the project that produces the error so you can check for yourselves.](https://github.com/godotengine/godot/files/2364165/String.zip) The idea is basically to solve the 1D wave equation using finite steps. I use a limited number of points (100). For now, the code simply makes a tiny step every certain time (so simulation goes veeery slow). I tried to first do the complete simulation and then simply update the position of the nodes in the `draw` method. No matter how I do this, it crashes. I believe the problem comes from the directive `draw_line`, which is called 100 times (one per node) each time I `update`.

It could be simply a problem with my graphics card driver, but I just want to be sure it is not a problem of the code generated by Godot.

**My setup:**
Godot 3.0.6 stable.
Ubuntu 16.04 LTS
OpenGL ES 3.0 Renderer: Mesa DRI Intel(R) HD Graphics 620 (Kabylake GT2)

**Comment by Noshyaar** [comment]
Can't reproduce with the given project. I adapted it a bit by yielding out of the simulation once in a while and it ran flawlessly. I don't believe it's because of `draw_line` but it could be that the simulation iterations took too long to complete in a frame.
```
Godot 3.0.6 stable
Linux 4.14.67-1-MANJARO x86_64 17.1.12 Hakoila
OpenGL ES 3.0 Renderer: Mesa DRI Intel(R) Haswell
```

**Comment by VictorSeven** [comment]
I think that it may be the problem -the draw code cannot complete in only one frame. I tried also as you did, making the whole simulation first and then updating the nodes, and still get the problem... I really don't know if there is a way to workaround this problem or if it is simply a problem of my graphics card.
Thank you for testing it!

**Comment by capnm** [comment]
Try to upgrade to Ubuntu 18.04 LTS, the Intel drivers have become much better.

**Comment by OvermindDL1** [comment]
> However, this time, my entire computer froze several times when running the app. Once my screen went corrupted, so I am pretty sure that the problem is that the graphic card froze.

This is not possible from a usermode application on linux (or that is a *major* kernel bug), this could only happen due to either a driver fault, system fault, or hardware fault.

Thus I concur with @capnm, you really need to upgrade to 18.04+, the older drivers from Intel were effectively garbage (as were a lot of the hardware from before that as well, they've gotten *MUCH* better now), their new drivers have been merged into the MESA system and kernel output and is of significantly better quality.  :-)

And hope there is no hardware fault of course.  ^.^;

**Comment by VictorSeven** [comment]
Thank you for your support!

I was able to test the project on another Ubuntu 16.04 with a (different) Intel graphics card and the project runs with no problem. So it really look like my computer has some kind of problem with the drivers. 

Simply one question more: do you think it would be possible to workaround the problem using this: https://github.com/godotengine/godot/pull/7269? 
I'm gonna implement that in any case, it looks like it will speed up a lot the simulation, since the drawing will happen only after X `steps` and not each frame.

**Comment by Chaosus** [comment]
Seems like a problem is purely driver related and it was fixed by the user, and in 3.1 the GLES2 mode is provided for such cases. I risk to close it..
